src/thumbnailer.cc

```diff
@@ -269,25 +314,91 @@ Thumbnailer::Status Thumbnailer::GenerateAnimationEqualPSNR(
   return SetLoopCount(webp_data);
 }
 
-Thumbnailer::Status Thumbnailer::SetLoopCount(WebPData* const webp_data) {
-  std::unique_ptr<WebPMux, void (*)(WebPMux*)> mux(WebPMuxCreate(webp_data, 1),
-                                                   WebPMuxDelete);
-  if (mux == nullptr) return kMemoryError;
+Thumbnailer::Status Thumbnailer::TryNearLossless(WebPData* const webp_data) {
+  int anim_size = webp_data->size;
 
-  WebPMuxAnimParams new_params;
-  if (WebPMuxGetAnimationParams(mux.get(), &new_params) != WEBP_MUX_OK) {
-    return kMemoryError;
+  WebPAnimEncoderDelete(enc_);
+  enc_ = WebPAnimEncoderNew(frames_[0].pic.width, frames_[0].pic.height,
+                            &anim_config_);
+  if (enc_ == nullptr) return kMemoryError;
+
+  std::cerr << std::endl
+            << "Final near-lossless's pre-processing values:" << std::endl;
+
+  int curr_index = 0;  // index of current frame.
+  for (auto& frame : frames_) {
+    int curr_size = results_[curr_index].size;
+    frame.config.quality = results_[curr_index].quality;
+    float curr_psnr = GetPSNR(&frame.pic, frame.config);
+
+    int min_near_ll = 0;
+    int max_near_ll = 100;
+    int final_near_ll = -1;
+
+    frame.config.lossless = 1;
+    frame.config.near_lossless = 0;
+    frame.config.quality = 90;
+    int new_size = GetSize(&frame.pic, frame.config);
+
+    // Only try binary search if near-lossles encoding with pre-processing = 0
+    // is feasible in order to save execution time.
+    if (anim_size - curr_size + new_size <= byte_budget_) {
+      // Binary search for near-lossless's pre-processing value.
+      while (min_near_ll <= max_near_ll) {
+        int mid_near_ll = (min_near_ll + max_near_ll) / 2;
+        frame.config.near_lossless = mid_near_ll;
+        new_size = GetSize(&frame.pic, frame.config);
+        if (anim_size - curr_size + new_size <= byte_budget_) {
+          float new_psnr = GetPSNR(&frame.pic, frame.config);
```

Comment:
For another pull request: encoding the image a second time just to get the distortion should be avoided.

Reply:
For later, how should we get the `PSNR` without ecoding the image a second time ?

Reply:
You could store the encoded bitstream of `GetSize()` and then decode+disto it only if needed.